langchain4j-core/src/main/java/dev/langchain4j/model/embedding/AbstractEmbeddingModel.java

```diff
@@ -0,0 +1,44 @@
+package dev.langchain4j.model.embedding;
+
+import java.util.Map;
+
+/**
+ * An Embedding Model which contains
+ */
+public abstract class AbstractEmbeddingModel implements EmbeddingModel {
+
+    /**
+     * embedding model dimension
+     */
+    protected Integer dimension;
+
+    /**
+     * A map contains known model's name and its dimension
+     *
+     * @return A map, key is the common represented model name, value is its dimension
+     */
+    protected abstract Map<String, Integer> dimensionMap();
```

Comment:
> The purpose to expose dimensionMap and modelName is just like EmbeddingStoreWithoutMetadataIT's embeddingModel() and embeddingStore(). I think it makes dimension() more standardized and uniform without having to worry about differences between modelName and dimensionMap in different implementations

Not sure I understand, can you please elaborate?

A couple of minor downsides that I see:
- `dimensionMap()` exposes much more information (dimensions for all models) than is required for the task. At some point someone will have a temptation to misuse it.
- having `Map` allows only matching models by exact name, some other implementation-specific operations like "modelName.startsWith("...")" are not possible

Reply:
The standard step to get dimension is:

1. check cached dimension, if exist, return it.
2. check known model's dimension, if there is any match, cache and return it.
3. manually embed "test" and cache it.

There are two differences in these steps:

1. different implementations may have different ways of getting model name, e.g. some implementations get it via the modelName attribute, some names may be, as you say, a prefix or something like that.
2. implementations may have different ways of storing known dimensional models, perhaps via enum, perhaps via map, etc., depending on the implementation.

But for `DimensionAwareEmbeddingModel`, all it has to do is to check if there is a corresponding cached dimension based on the modelName, and if there is one, get it directly, if not, embed it manually and get it. So I think their exposure is necessary.

> dimensionMap() exposes much more information (dimensions for all models) than is required for the task. At some point someone will have a temptation to misuse it.

Do you have any good ideas to avoid it? I currently have it set to `protected`, so it is invisible to users. Maybe we should rename it to `getKnownDimension()` to avoid misuse?

> having Map allows only matching models by exact name, some other implementation-specific operations like "modelName.startsWith("...")" are not possible

That's a real downside...Maybe we need a `Function` to calculate the dimension, like `getKnownDimension()`. This allows implementations to implement it more freely.

So the solution may like below:

```java
public int dimension() {
    if (dimension != null) {
        return dimension;
    }
    Integer knownDimension = getKnownDimension();  // different implementations have different ways to get known dimension
    dimension = Optional.ofNullable(knownDimension).orElseGet(() -> embed("test").content().dimension());
    return dimension;
}

Reply:
```java
public int dimension() {
    if (dimension != null) {
        return dimension;
    }
    Integer knownDimension = getKnownDimension();  // different implementations have different ways to get known dimension
    dimension = Optional.ofNullable(knownDimension).orElseGet(() -> embed("test").content().dimension());
    return dimension;
}
```

Maybe this will make implementations to get known dimension more freely